### PR TITLE
Using canonical etype instead of etype in distributed partition

### PR DESCRIPTION
**Pull request by VoVAllen**
## Description
Fix https://github.com/dmlc/dgl/issues/3247

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/distributed/partition.py [hunk 1]
```diff
@@ -42,6 +44,49 @@ def _get_part_ranges(id_ranges):
...
+    copy_d = copy.deepcopy(part_d)
+    # Iterate over d, modify on copy_d
+
+    def _recursive_convert_str_to_tuple_key(part_d, key_tuple=tuple()):
+        if isinstance(part_d, dict):
+            for k, v in part_d.items():
+                sub_d = copy_d
+                for key in key_tuple:
+                    sub_d = sub_d[key]
+                if k.startswith("tuple"):
```
Comment (BarclayII) [inline python/dgl/distributed/partition.py:58]: What if someone names their edge type "tuple" (though unlikely)?

Is it possible to always encode the canonical edge types in JSON?

Comment (BarclayII) [reply python/dgl/distributed/partition.py:58]: NVM the edge type keys are always replaced with canonical edge types so this is not a problem.

File: python/dgl/distributed/partition.py [hunk 2]
```diff
@@ -42,6 +44,49 @@ def _get_part_ranges(id_ranges):
...
+                        sub_d.pop(k)
+                    except ValueError:
+                        print("Error parse: {}".format(k))
+                _recursive_convert_str_to_tuple_key(v, key_tuple+(k,))
+        # return copy_d
+    _recursive_convert_str_to_tuple_key(part_d)
+    return copy_d
+
+
+def _convert_tuple_key_to_str(part_d):
```
Comment (BarclayII) [inline python/dgl/distributed/partition.py:71]: Please add a comment (or docstring?) that the edge type keys should always be in canonical form.

**Review by BarclayII — APPROVED** [review]

**Comment by VoVAllen** [comment]
Found this is much more complex than I thought since handling canonical etype/etype is not designed at the beginning. Decided to postpone this feature request

**Comment by sbocconi** [comment]
The problem applies also to `HeteroGraphConv` which has as input:
```
mods : dict[str, nn.Module]
        Modules associated with every edge types. 
```
How to use it in case the same relation is used between different types of nodes?

I get the same error trying to train this kind of graphs:
```
metagraph=[('tweet', 'url', 'links'), ('tweet', 'user', 'mentions'), ('tweet', 'hashtag', 'tags'), ('user', 'url', 'links'), ('user', 'user', 'mentions'), ('user', 'hashtag', 'tags'), ('user', 'tweet', 'tweets')])
```
```
dgl._ffi.base.DGLError: Edge type "links" is ambiguous. Please use canonical edge type in the form of (srctype, etype, dsttype)
```

**Pull request closed without merging**